beacon_node/http_api/src/lib.rs

```diff
@@ -402,40 +403,71 @@ pub fn serve<T: BeaconChainTypes>(
             })
         });
 
-    // GET beacon/states/{state_id}/validators
+    // GET beacon/states/{state_id}/validators?id,status
     let get_beacon_state_validators = beacon_states_path
         .clone()
         .and(warp::path("validators"))
+        .and(warp::query::<api_types::ValidatorsQuery>())
         .and(warp::path::end())
-        .and_then(|state_id: StateId, chain: Arc<BeaconChain<T>>| {
-            blocking_json_task(move || {
-                state_id
-                    .map_state(&chain, |state| {
-                        let epoch = state.current_epoch();
-                        let finalized_epoch = state.finalized_checkpoint.epoch;
-                        let far_future_epoch = chain.spec.far_future_epoch;
-
-                        Ok(state
-                            .validators
-                            .iter()
-                            .zip(state.balances.iter())
-                            .enumerate()
-                            .map(|(index, (validator, balance))| api_types::ValidatorData {
-                                index: index as u64,
-                                balance: *balance,
-                                status: api_types::ValidatorStatus::from_validator(
-                                    Some(validator),
-                                    epoch,
-                                    finalized_epoch,
-                                    far_future_epoch,
-                                ),
-                                validator: validator.clone(),
-                            })
-                            .collect::<Vec<_>>())
-                    })
-                    .map(api_types::GenericResponse::from)
-            })
-        });
+        .and_then(
+            |state_id: StateId, chain: Arc<BeaconChain<T>>, query: api_types::ValidatorsQuery| {
+                blocking_json_task(move || {
+                    let id_predicate = |index: &u64, validator: &Validator| match query.id.as_ref()
```

Comment:
With the use of [`map_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) you might be able to avoid these predicates.

Reply:
Let me know if I got what you were thinking